### PR TITLE
Introduce delete[edge]Waypoint.spec.ts

### PR DESCRIPTION
**Pull request by jomarko**
In this PR we introduce two kind of test:
- `deleteEdgeWaypoint`  with checking the DOM structure
- `moveAndDeleteEdgeWaypoint` with checking visual effect of the waypoint deletion

### CODE REVIEW

File: packages/dmn-editor/tests/e2e/drdArtifacts/deleteAssociationWaypoint.spec.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      targetPosition: { x: 500, y: 100 },
+    });
+    await edges.moveNthWaypoint({
+      from: DefaultNodeName.INPUT_DATA,
+      to: DefaultNodeName.TEXT_ANNOTATION,
+      nth: 2,
+      targetPosition: { x: 500, y: 500 },
+    });
+
+    await expect(diagram.get()).toHaveScreenshot("add-multiple-association-waypoint-and-move-them.png");
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/drdArtifacts/deleteAssociationWaypoint.spec.ts:82]: needs to be new file

File: packages/dmn-editor/tests/e2e/drgRequirements/deleteAuthorityRequirementWaypoint.spec.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  test("should delete the single Association edge waypoint to make it straight", async ({ diagram, edges }) => {
+    await edges.addWaypoint({ from: DefaultNodeName.INPUT_DATA, to: DefaultNodeName.KNOWLEDGE_SOURCE });
+    await edges.moveNthWaypoint({
+      from: DefaultNodeName.INPUT_DATA,
+      to: DefaultNodeName.KNOWLEDGE_SOURCE,
+      nth: 1,
+      targetPosition: { x: 300, y: 300 },
+    });
+
+    await expect(diagram.get()).toHaveScreenshot("add-authority-requirement-waypoint-and-move-it.png");
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/drgRequirements/deleteAuthorityRequirementWaypoint.spec.ts:54]: needs to be new file

File: packages/dmn-editor/tests/e2e/drgRequirements/deleteAuthorityRequirementWaypoint.spec.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      targetPosition: { x: 500, y: 100 },
+    });
+    await edges.moveNthWaypoint({
+      from: DefaultNodeName.INPUT_DATA,
+      to: DefaultNodeName.KNOWLEDGE_SOURCE,
+      nth: 2,
+      targetPosition: { x: 500, y: 500 },
+    });
+
+    await expect(diagram.get()).toHaveScreenshot("add-multiple-authority-requirement-waypoints-and-move-them.png");
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/drgRequirements/deleteAuthorityRequirementWaypoint.spec.ts:83]: needs to be new file

File: packages/dmn-editor/tests/e2e/drgRequirements/deleteInformationRequirementWaypoint.spec.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  }) => {
+    await edges.addWaypoint({ from: DefaultNodeName.INPUT_DATA, to: DefaultNodeName.DECISION });
+    await edges.moveNthWaypoint({
+      from: DefaultNodeName.INPUT_DATA,
+      to: DefaultNodeName.DECISION,
+      nth: 1,
+      targetPosition: { x: 300, y: 300 },
+    });
+
+    await expect(diagram.get()).toHaveScreenshot("add-information-requirement-waypoint-and-move-it.png");
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/drgRequirements/deleteInformationRequirementWaypoint.spec.ts:57]: needs to be new file

File: packages/dmn-editor/tests/e2e/drgRequirements/deleteInformationRequirementWaypoint.spec.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      targetPosition: { x: 500, y: 100 },
+    });
+    await edges.moveNthWaypoint({
+      from: DefaultNodeName.INPUT_DATA,
+      to: DefaultNodeName.DECISION,
+      nth: 2,
+      targetPosition: { x: 500, y: 500 },
+    });
+
+    await expect(diagram.get()).toHaveScreenshot("add-multiple-information-requirement-waypoints-and-move-them.png");
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/drgRequirements/deleteInformationRequirementWaypoint.spec.ts:87]: needs to be new file

File: packages/dmn-editor/tests/e2e/drgRequirements/deleteKnowledgeRequirementWaypoint.spec.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  }) => {
+    await edges.addWaypoint({ from: DefaultNodeName.BKM, to: DefaultNodeName.DECISION });
+    await edges.moveNthWaypoint({
+      from: DefaultNodeName.BKM,
+      to: DefaultNodeName.DECISION,
+      nth: 1,
+      targetPosition: { x: 300, y: 300 },
+    });
+
+    await expect(diagram.get()).toHaveScreenshot("add-knowledge-requirement-waypoint-and-move-it.png");
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/drgRequirements/deleteKnowledgeRequirementWaypoint.spec.ts:57]: needs to be new file

File: packages/dmn-editor/tests/e2e/drgRequirements/deleteKnowledgeRequirementWaypoint.spec.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+      targetPosition: { x: 500, y: 100 },
+    });
+    await edges.moveNthWaypoint({
+      from: DefaultNodeName.BKM,
+      to: DefaultNodeName.DECISION,
+      nth: 2,
+      targetPosition: { x: 500, y: 500 },
+    });
+
+    await expect(diagram.get()).toHaveScreenshot("add-multiple-knowledge-requirement-waypoints-and-not-move-them.png");
```
Comment (jomarko) [inline packages/dmn-editor/tests/e2e/drgRequirements/deleteKnowledgeRequirementWaypoint.spec.ts:91]: needs to be new file

**Review by jomarko — COMMENTED** [review]
As discussed in separate Slack conversation, we can not reuse screenshots,

File: packages/dmn-editor/tests/e2e/drdArtifacts/deleteAssociationWaypoint.spec.ts [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+      from: DefaultNodeName.INPUT_DATA,
+      to: DefaultNodeName.TEXT_ANNOTATION,
+      waypointIndex: 1,
+    });
+    await expect(diagram.get()).toHaveScreenshot("delete-association-waypoint-straight-edge.png");
+  });
+
+  test("should delete all Association edge waypoints", async ({ nodes, edges }) => {
+    await edges.addWaypoint({ from: DefaultNodeName.INPUT_DATA, to: DefaultNodeName.TEXT_ANNOTATION });
+    await nodes.move({ name: DefaultNodeName.TEXT_ANNOTATION, targetPosition: { x: 200, y: 500 } });
```
Comment (ljmotta) [inline packages/dmn-editor/tests/e2e/drdArtifacts/deleteAssociationWaypoint.spec.ts:82]: please, use the move waypoint.

File: packages/dmn-editor/tests/e2e/drdArtifacts/deleteAssociationWaypoint.spec.ts [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+    await edges.addWaypoint({ from: DefaultNodeName.INPUT_DATA, to: DefaultNodeName.TEXT_ANNOTATION });
```
Comment (ljmotta) [inline packages/dmn-editor/tests/e2e/drdArtifacts/deleteAssociationWaypoint.spec.ts:84]: As the "GIVEN" part is very simple, I guess we could group them instead of breaking a new line.

File: packages/dmn-editor/tests/e2e/drgRequirements/deleteAuthorityRequirementWaypoint.spec.ts [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+    await palette.dragNewNode({ type: NodeType.INPUT_DATA, targetPosition: { x: 100, y: 100 } });
+    await nodes.dragNewConnectedNode({
+      from: DefaultNodeName.INPUT_DATA,
+      type: NodeType.KNOWLEDGE_SOURCE,
+      targetPosition: { x: 100, y: 300 },
+    });
+  });
+
+  test("should delete the single Association edge waypoint", async ({ edges }) => {
```
Comment (ljmotta) [inline packages/dmn-editor/tests/e2e/drgRequirements/deleteAuthorityRequirementWaypoint.spec.ts:45]:
```suggestion
  test("should delete the single Authority Requirement edge waypoint", async ({ edges }) => {
```

File: packages/dmn-editor/tests/e2e/drgRequirements/deleteAuthorityRequirementWaypoint.spec.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    await expect(
+      await edges.getWaypoint({
+        from: DefaultNodeName.INPUT_DATA,
+        to: DefaultNodeName.KNOWLEDGE_SOURCE,
+        waypointIndex: 1,
+      })
+    ).not.toBeAttached();
+  });
+
+  test("should delete the single Association edge waypoint to make it straight", async ({ diagram, edges }) => {
```
Comment (ljmotta) [inline packages/dmn-editor/tests/e2e/drgRequirements/deleteAuthorityRequirementWaypoint.spec.ts:63]:
```suggestion
  test("should delete the single Authority Requirement edge waypoint to make it straight", async ({ diagram, edges }) => {
```

**Review by ljmotta — CHANGES_REQUESTED** [review]
Regarding the test names, sometimes you use `single` others `one`. Please consider this as the test names as we don't need to specify what is the outcome of deleting a waypoint.

`should create and delete one Association edge waypoint`
`should create and delete two Association edge waypoints`
`should create two Association edge waypoints and delete one`

File: packages/dmn-editor/tests/e2e/drdArtifacts/deleteAssociationWaypoint.spec.ts (hunk 9 above)
Comment (jomarko) [reply packages/dmn-editor/tests/e2e/drdArtifacts/deleteAssociationWaypoint.spec.ts:84]: yes, sounds good, will do for all "should delete all ..."

File: packages/dmn-editor/tests/e2e/drdArtifacts/deleteAssociationWaypoint.spec.ts (hunk 8 above)
Comment (jomarko) [reply packages/dmn-editor/tests/e2e/drdArtifacts/deleteAssociationWaypoint.spec.ts:82]: here I think I can not incorporate your suggestion, currently the only way to have two or more waypoints is to:
- add first waypoint
- move one ending node, but the middle of the edge needs to be the same as middle between nodes, as adding waypoint simply click into the middle of the edge bounding box
- add next waypoint
- move ending node in the same manner as previously
- ...

Comment (ljmotta) [reply packages/dmn-editor/tests/e2e/drdArtifacts/deleteAssociationWaypoint.spec.ts:82]: Ok, maybe I didn't paid attention to this detail on the other tests. But as I see now, the `{ x: 200, y: 500 }` is not an arbitrary value. I don't think this is a good approach as it works in a "mysterious" way. I mean, we don't have a comment explaining it, and we need to think that maybe we'll not remember this in a couple of months. We could add a very good comment for our future selfs on all tests, or we could change this so we can click on the middle of an edge segment whenever we want.

On the `path` element we have the `d` attribute, which is a string with the edge points. `M` as the starting point and `L` as the ending point. For an edge with multiple waypoints it will have as many `M` and `L` pairs, creating the edge segments. Our Diagram has an offset of `100,100`, which means if the point marks `100,100` it's actually `200,200`. Parsing the string, and knowing the two points `M` and `L` can tell exactly where is the middle point of the edge segment. For example, an edge with one waypoint can have the following `d` attribute `M 140,140 L 260,80 M 260,80 L 380,140`. If we want to click on the middle of the first edge segment: 

```
M 140, 140 (starting)
L 260, 80 (ending)

calculate the middle point
400, 240 (sum)
200, 120 (divide by 2)
300, 240 (add offset)
```

Clicking at `300,240` should click at the middle of the first edge segment if we don't have any other offset to apply. Now, this could be translated to a method on the edge fixture. `addWaypointBetweenConsecutiveWaypoints(args: { from: string, to: string, startingWaypoint: number  })`. I don't think `startingWaypoint` is a good name, but I couldn't think on a better one. The `startingWaypoint` would give us which pair we need to use from the `d` string.

**Comment by jomarko** [comment]
Incorporating feedback about add waypoint fixture

**Review by ljmotta — APPROVED** [review]
